### ISSUE TITLE
Can I train your codes with 4 Gpus?

### ISSUE THREAD

**Issue by longmalongma** [hkchengrex/STCN#62]
Can I train code with 4 Gpus? What changes need to be made for the training of 4 Gpus? Also, should 4 Gpus train faster? But will the segmentation accuracy be reduced?

**Comment by hkchengrex** [comment]
In the training command, change `CUDA_VISIBLE_DEVICES=[a,b]` to `CUDA_VISIBLE_DEVICES=[a,b,c,d]`  where a, b, c, d are the four GPU ids.
And change `--nproc_per_node=2` to `--nproc_per_node=4`.
The total effective batch size is the batch size specified (default in https://github.com/hkchengrex/STCN/blob/main/util/hyper_para.py) multiplied by the total number of processes. This can affect performance so you might want to tune it.

**Comment by longmalongma** [comment]
> In the training command, change `CUDA_VISIBLE_DEVICES=[a,b]` to `CUDA_VISIBLE_DEVICES=[a,b,c,d]` where a, b, c, d are the four GPU ids.
> And change `--nproc_per_node=2` to `--nproc_per_node=4`.
> The total effective batch size is the batch size specified (default in https://github.com/hkchengrex/STCN/blob/main/util/hyper_para.py) multiplied by the total number of processes (in your case, 4). This can affect performance so you might want to tune it.

@hkchengrex Ok，thnaks for your reply. Can running on a 4GPU save 1/2 time in both pre-training and main training?For 4GPU training, do you have any suggestions on the number of batch size to be set during the three-stage training

**Comment by hkchengrex** [comment]
That depends on your hardware.
You can try to either keep the same total batch size or double the total batch size while reducing the total number of iterations.

**Comment by longmalongma** [comment]
> That depends on your hardware.
> You can try to either keep the same total batch size or double the total batch size while reducing the total number of iterations.
If I use 4GPU, I need to set the Batch size to 8 and iterations to 150,000?Is that so?
![image](https://user-images.githubusercontent.com/40291722/134833599-d89e0853-21f0-4624-914d-0b3fa1ff862a.png)

**Comment by longmalongma** [comment]
> That depends on your hardware.
> You can try to either keep the same total batch size or double the total batch size while reducing the total number of iterations.

@hkchengrex When using 4 Gpus, other parameters remain unchanged except batch size and interations, right?

**Comment by hkchengrex** [comment]
Sometimes people also scale the learning rate by the same factor when they increase the batch size. I don't really know an "optimal" hyperparameter setting, nor is the original setting "optimal".
Limited hyperparameter search was performed due to computational constraints.

**Comment by longmalongma** [comment]
> Sometimes people also scale the learning rate by the same factor when they increase the batch size. I don't really know an "optimal" hyperparameter setting, nor is the original setting "optimal".
> Limited hyperparameter search was performed due to computational constraints.

Ok, thanks!

**Comment by zhouweii234** [comment]
@longmalongma May I ask if the results of training with 4 GPUs are the same as in the article ？
